File: src/test/java/pl/mjedynak/idea/plugins/builder/gui/displayer/GenerateBuilderPopupDisplayerTest.java
package pl.mjedynak.idea.plugins.builder.gui.displayer;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.ui.popup.JBPopup;
import com.intellij.openapi.ui.popup.PopupChooserBuilder;
import javax.swing.JList;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockedStatic;
import org.mockito.junit.jupiter.MockitoExtension;
import pl.mjedynak.idea.plugins.builder.factory.PopupChooserBuilderFactory;

@ExtendWith(MockitoExtension.class)
public class GenerateBuilderPopupDisplayerTest {

    private static final String TITLE = "Builder already exists";

    @InjectMocks
    private GenerateBuilderPopupDisplayer popupDisplayer;

    @Mock
    private MockedStatic<PopupChooserBuilderFactory> popupChooserBuilderFactory;

    @Mock
    private JList<?> list;

    @SuppressWarnings("rawtypes")
    @Mock
    private PopupChooserBuilder popupChooserBuilder;

    @Mock
    private Editor editor;

    @Mock
    private JBPopup popup;

    @Test
    void shouldInvokePopupChooserBuilder() {
        // given
        popupChooserBuilderFactory
                .when(() -> PopupChooserBuilderFactory.getPopupChooserBuilder(list))
                .thenReturn(popupChooserBuilder);
        given(popupChooserBuilder.setTitle(TITLE)).willReturn(popupChooserBuilder);
        given(popupChooserBuilder.setItemChosenCallback(any(Runnable.class))).willReturn(popupChooserBuilder);
        given(popupChooserBuilder.setMovable(true)).willReturn(popupChooserBuilder);
        given(popupChooserBuilder.createPopup()).willReturn(popup);

        // when
        popupDisplayer.displayPopupChooser(editor, list, () -> {});

        // then
        verify(popup).showInBestPositionFor(editor);
    }
}

File: src/test/kotlin/io/github/kingg22/kobra/builder/gui/displayer/GenerateBuilderPopupDisplayerTest.kt
package io.github.kingg22.kobra.builder.gui.displayer

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.ui.popup.JBPopup
import com.intellij.openapi.ui.popup.PopupChooserBuilder
import io.github.kingg22.kobra.builder.factory.PopupChooserBuilderFactory
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.MockedStatic
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import javax.swing.JList

@ExtendWith(MockitoExtension::class)
class GenerateBuilderPopupDisplayerTest {
    @InjectMocks
    private lateinit var popupDisplayer: GenerateBuilderPopupDisplayer

    @Mock
    private lateinit var popupChooserBuilderFactory: MockedStatic<PopupChooserBuilderFactory>

    @Mock
    private lateinit var list: JList<*>

    @Mock
    private lateinit var popupChooserBuilder: PopupChooserBuilder<*>

    @Mock
    private lateinit var editor: Editor

    @Mock
    private lateinit var popup: JBPopup

    @Test
    fun shouldInvokePopupChooserBuilder() {
        // given
        popupChooserBuilderFactory
            .`when`<PopupChooserBuilder<*>> { PopupChooserBuilderFactory.getPopupChooserBuilder(list) }
            .thenReturn(popupChooserBuilder)
        given(popupChooserBuilder.setTitle(TITLE))
            .willReturn(popupChooserBuilder)
        given(popupChooserBuilder.setItemChosenCallback(any<Runnable>()))
            .willReturn(popupChooserBuilder)
        given(popupChooserBuilder.setMovable(true))
            .willReturn(popupChooserBuilder)
        given(popupChooserBuilder.createPopup()).willReturn(popup)

        // when
        popupDisplayer.displayPopupChooser(editor, list) {}

        verify(popup).showInBestPositionFor(editor)
    }

    companion object {
        private const val TITLE = "Builder already exists"
    }
}
